codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,297 @@
+        if (pState.isActive) {
+            int nsent = min(maxBytes, pState.streamableBytes);
+            pState.write(nsent, writer);
+            if (nsent == 0) {
+                // If a stream sends zero bytes, then we gave it a chance to write empty frames and it is now
+                // considered inactive until the next call to updateStreamableBytes. This allows descendant streams to
+                // be allocated bytes when the parent stream can't utilize them. This may be as a result of the
+                // stream's flow control window being 0.
+                pState.updateStreamableBytes(pState.streamableBytes, false);
+                return 0;
+            }
+            connectionState.pseudoTime += nsent;
+            return nsent;
+        }
+
+        return distributeToChildren(maxBytes, writer, pState);
+    }
+
+    private int distributeToChildren(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        long oldTotalQueuedWeights = pState.totalQueuedWeights;
+        State iState = pState.poll();
+        State iNextState = pState.peek();
+        try {
+            assert iNextState == null || iNextState.pseudoTime >= iState.pseudoTime : "iNextState.pseudoTime(" +
+                    iNextState.pseudoTime + ") < " + " iState.pseudoTime(" + iState.pseudoTime + ")";
+            int nsent = distribute(iNextState == null ? maxBytes :
+                                    min(maxBytes, (int) ((iNextState.pseudoTime - iState.pseudoTime) *
+                                             ((double) iNextState.stream.weight() / oldTotalQueuedWeights)
+                                             + allocationQuantum)),
+                               writer,
+                               iState);
+            iState.updatePseudoTime(nsent, oldTotalQueuedWeights);
+            return nsent;
+        } finally { // Do in finally to ensure the internal state is not corrupted if an exception is thrown.
+            // The offer operation is delayed until we unroll up the recursive stack, so we don't have to remove from
+            // the priority queue due to a write operation.
+            if (iState.isActiveCountForTree != 0) {
+                pState.offer(iState);
+            }
+        }
+    }
+
+    private State state(Http2Stream stream) {
+        return stream.getProperty(stateKey);
+    }
+
+    /**
+     * For testing only!
+     */
+    int streamableBytes(Http2Stream stream) {
+        return state(stream).streamableBytes;
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State implements PriorityQueueNode<State> {
+        final Http2Stream stream;
+        private final Queue<State> queue;
+        int streamableBytes;
+        int isActiveCountForTree;
+        private int priorityQueueIndex = INDEX_NOT_IN_QUEUE;
+        long pseudoTime;
```

Comment:
This psuedoTime is used for two different things: the "current time" and the time this stream should be sent.  Unfortunately, each parent should have its own individual progression of time (can't just use the connection's `psuedoTime`), independent of what is sent in other parts of the tree. That's because the time-scaling used when a child sends is based only on the immediate children of its parent.